instrumentation/net/http/otelhttp/transport.go

```diff
@@ -109,21 +143,69 @@ func (t *Transport) RoundTrip(r *http.Request) (*http.Response, error) {
 		ctx = httptrace.WithClientTrace(ctx, t.clientTrace(ctx))
 	}
 
+	readRecordFunc := func(int64) {}
+	if t.readEvent {
+		readRecordFunc = func(n int64) {
+			span.AddEvent("read", trace.WithAttributes(ReadBytesKey.Int64(n)))
+		}
+	}
+
+	var bw bodyWrapper
+	// if request body is nil or NoBody, we don't want to mutate the body as it
+	// will affect the identity of it in an unforeseeable way because we assert
+	// ReadCloser fulfills a certain interface and it is indeed nil or NoBody.
+	if r.Body != nil && r.Body != http.NoBody {
+		bw.ReadCloser = r.Body
+		bw.record = readRecordFunc
+		r.Body = &bw
+	}
+
 	r = r.WithContext(ctx)
 	span.SetAttributes(httpconv.ClientRequest(r)...)
+	if t.getRequestAttributes != nil {
+		span.SetAttributes(t.getRequestAttributes(r)...)
+	}
 	t.propagators.Inject(ctx, propagation.HeaderCarrier(r.Header))
 
 	res, err := t.rt.RoundTrip(r)
 	if err != nil {
 		span.RecordError(err)
 		span.SetStatus(codes.Error, err.Error())
 		span.End()
-		return res, err
+	} else {
+		span.SetAttributes(httpconv.ClientResponse(res)...)
+		if t.getResponseAttributes != nil {
+			span.SetAttributes(t.getResponseAttributes(res)...)
+		}
+		span.SetStatus(httpconv.ClientStatus(res.StatusCode))
+		res.Body = newWrappedBody(span, res.Body)
+	}
+
+	// Add metrics
+	attributes := httpconv.ClientRequest(r)
+	if t.getRequestAttributes != nil {
+		attributes = append(attributes, t.getRequestAttributes(r)...)
 	}
+	if err == nil {
+		if res.StatusCode > 0 {
+			attributes = append(attributes, semconv.HTTPStatusCode(res.StatusCode))
+		}
+		if t.getResponseAttributes != nil {
+			attributes = append(attributes, t.getResponseAttributes(res)...)
+		}
+	} else {
+		attributes = append(attributes, semconv.HTTPStatusCode(http.StatusBadRequest))
+	}
+	o := metric.WithAttributes(attributes...)
+	t.counters[ClientRequestContentLength].Add(ctx, bw.read, o)
```

Comment:
Hmm this is the [same code as the middleware one](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/instrumentation/net/http/otelhttp/handler.go#L222), wonder why this isn't triggered there?

Reply:
At quick glance two possibilities

1. The code in https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/instrumentation/net/http/otelhttp/handler.go#L222 runs in a single goroutine therefore does not need synchronizing 
2. The probability of hitting this race is very low 😉

Reply:
> does code in `Transport` runs in more than one Goroutine? 

Yes. You can find the stacktrace here: https://github.com/open-telemetry/opentelemetry-go-contrib/actions/runs/4883405160/jobs/8714784942

> Did you understand what the small probably is?

Guessing 1 / 100 000 on my machine 😉

Reply:
```
rpajak@LAPTOP-K6FOL8LO:~/repos/opentelemetry-go-contrib/instrumentation/net/http/otelhttp/test$ go test -race -count=100000
==================
WARNING: DATA RACE
Write at 0x00c01cb7fe78 by goroutine 2289465:
  go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.(*bodyWrapper).Read()
      /home/rpajak/repos/opentelemetry-go-contrib/instrumentation/net/http/otelhttp/wrap.go:40 +0xa4
  io.discard.ReadFrom()
      /usr/local/go/src/io/io.go:651 +0x91
  io.(*discard).ReadFrom()
      <autogenerated>:1 +0x5d
  io.copyBuffer()
      /usr/local/go/src/io/io.go:413 +0x1c5
  io.Copy()
      /usr/local/go/src/io/io.go:386 +0x64
  net/http.(*transferWriter).doBodyCopy()
      /usr/local/go/src/net/http/transfer.go:412 +0x3e
  net/http.(*transferWriter).writeBody()
      /usr/local/go/src/net/http/transfer.go:375 +0x804
  net/http.(*Request).write()
      /usr/local/go/src/net/http/request.go:705 +0x11dd
  net/http.(*persistConn).writeLoop()
      /usr/local/go/src/net/http/transport.go:2413 +0x2e4
  net/http.(*Transport).dialConn.func6()
      /usr/local/go/src/net/http/transport.go:1766 +0x39

Previous read at 0x00c01cb7fe78 by goroutine 2289453:
  go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.(*Transport).RoundTrip()
      /home/rpajak/repos/opentelemetry-go-contrib/instrumentation/net/http/otelhttp/transport.go:200 +0x18de
  net/http.send()
      /usr/local/go/src/net/http/client.go:252 +0x942
  net/http.(*Client).send()
      /usr/local/go/src/net/http/client.go:176 +0x164
  net/http.(*Client).do()
      /usr/local/go/src/net/http/client.go:716 +0x10de
  net/http.(*Client).Do()
      /usr/local/go/src/net/http/client.go:582 +0x1c4
  go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.Post()
      /home/rpajak/repos/opentelemetry-go-contrib/instrumentation/net/http/otelhttp/client.go:55 +0x1a5
  go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp/test.TestConvenienceWrappers()
      /home/rpajak/repos/opentelemetry-go-contrib/instrumentation/net/http/otelhttp/test/client_test.go:67 +0x76c
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1629 +0x47

Goroutine 2289465 (running) created at:
  net/http.(*Transport).dialConn()
      /usr/local/go/src/net/http/transport.go:1766 +0x2619
  net/http.(*Transport).dialConnFor()
      /usr/local/go/src/net/http/transport.go:1456 +0x13a
  net/http.(*Transport).queueForDial.func1()
      /usr/local/go/src/net/http/transport.go:1425 +0x47

Goroutine 2289453 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1629 +0x805
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:2036 +0x8d
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x216
  testing.runTests()
      /usr/local/go/src/testing/testing.go:2034 +0x87c
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1906 +0xb44
  main.main()
      _testmain.go:77 +0x2e9
==================
--- FAIL: TestConvenienceWrappers (0.00s)
    testing.go:1446: race detected during execution of test
panic: test timed out after 10m0s
running tests:
        TestWithHTTPTrace (0s)

goroutine 3293126 [running]:
testing.(*M).startAlarm.func1()
        /usr/local/go/src/testing/testing.go:2241 +0x219
created by time.goFunc
        /usr/local/go/src/time/sleep.go:176 +0x48

goroutine 1 [chan receive]:
testing.(*T).Run(0xc00904d520, {0x9ff136, 0x11}, 0xa2fe70)
        /usr/local/go/src/testing/testing.go:1630 +0x82e
testing.runTests.func1(0x0?)
        /usr/local/go/src/testing/testing.go:2036 +0x8e
testing.tRunner(0xc00904d520, 0xc003671b48)
        /usr/local/go/src/testing/testing.go:1576 +0x217
testing.runTests(0xc00014bae0?, {0xd32f40, 0x10, 0x10}, {0x1c?, 0x1e?, 0xd3db20?})
        /usr/local/go/src/testing/testing.go:2034 +0x87d
testing.(*M).Run(0xc00014bae0)
        /usr/local/go/src/testing/testing.go:1906 +0xb45
main.main()
        _testmain.go:77 +0x2ea

goroutine 3293139 [select]:
net/http.(*Transport).getConn(0xd32c80, 0xc004adf600, {{}, 0x0, {0xc00228ef48, 0x4}, {0xc00d927861, 0xf}, 0x0})
        /usr/local/go/src/net/http/transport.go:1382 +0xa7c
net/http.(*Transport).roundTrip(0xd32c80, 0xc00988ad00)
        /usr/local/go/src/net/http/transport.go:590 +0xcff
net/http.(*Transport).RoundTrip(0xc010da9200?, 0xacd360?)
        /usr/local/go/src/net/http/roundtrip.go:17 +0x37
go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.(*Transport).RoundTrip(0xc006b6c370, 0xc00988ac00)
        /home/rpajak/repos/opentelemetry-go-contrib/instrumentation/net/http/otelhttp/transport.go:170 +0xc05
net/http.send(0xc00988ac00, {0xac8e60, 0xc006b6c370}, {0x8?, 0xc000100000?, 0x0?})
        /usr/local/go/src/net/http/client.go:252 +0x943
net/http.(*Client).send(0xc00ee30570, 0xc00988ac00, {0x0?, 0x296000?, 0x0?})
        /usr/local/go/src/net/http/client.go:176 +0x165
net/http.(*Client).do(0xc00ee30570, 0xc00988ac00)
        /usr/local/go/src/net/http/client.go:716 +0x10df
net/http.(*Client).Do(...)
        /usr/local/go/src/net/http/client.go:582
go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp/test.TestWithHTTPTrace(0xc009ad3a00)
        /home/rpajak/repos/opentelemetry-go-contrib/instrumentation/net/http/otelhttp/test/transport_test.go:284 +0x665
testing.tRunner(0xc009ad3a00, 0xa2fe70)
        /usr/local/go/src/testing/testing.go:1576 +0x217
created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:1629 +0x806

goroutine 3292992 [runnable]:
net/http/httptest.(*Server).goServe.func1()
        /usr/local/go/src/net/http/httptest/server.go:308
created by net/http/httptest.(*Server).goServe
        /usr/local/go/src/net/http/httptest/server.go:308 +0xa7

goroutine 3292993 [runnable]:
net/http.(*Transport).queueForDial.func1()
        /usr/local/go/src/net/http/transport.go:1425
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:1598 +0x1
created by net/http.(*Transport).queueForDial
        /usr/local/go/src/net/http/transport.go:1425 +0x17c
exit status 2
FAIL    go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp/test      600.119s
```

Reply:
Looks like the only fix will be to add a mutex to `bodyWrapper` then. It is only used in tests anyway.

Reply:
> It is only used in tests anyway.

No. It is not only used in tests.

> Looks like the only fix will be to add a mutex

You can use https://pkg.go.dev/sync/atomic#Int64 to avoid locking.